### PR TITLE
Implement proper support for nested complex env values

### LINKED ISSUE #2304: Environment names for complex types
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this feature/change is needed
* [x] After submitting this, I commit to one of:
    * Look through open issues and helped at least one other person
    * Hit the "watch" button on this repo to receive notifications and I commit to help at least 2 people that ask questions in the future
    * Implement a Pull Request for a confirmed bug

# Feature Request

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7.3
            pydantic compiled: True
                 install path: /tmp/pydantic-compex-env-names/env/lib/python3.9/site-packages/pydantic
               python version: 3.9.1 (default, Dec 13 2020, 11:55:53)  [GCC 10.2.0]
                     platform: Linux-5.10.7-3-MANJARO-x86_64-with-glibc2.32
     optional deps. installed: []
```

```py
from typing import Dict
from pydantic import BaseModel, BaseSettings, Field

class Backend(BaseModel):
    type: str = None
    dsn: str = None

class Settings(BaseSettings):
    backends: Dict[str, Backend] = Field(default_factory=dict)
```

In order to manage these settings via envvars, nested components could be separated with `__`, like this:

```sh
BACKENDS__DEFAULT__TYPE=postgresql
BACKENDS__DEFAULT__DSN=postgresql://localhost/db
```

In order to remove `default` backend, following envvar could be set:

```sh
BACKENDS__DEFAULT=
```

Or list of dict keys could be given like this:

```sh
BACKENDS=default,sqlite
```

This would remove all backends, except `default` and `sqlite`.

I found similar [issue](https://github.com/samuelcolvin/pydantic/issues/1458) about adding custom value parsers, so this could be used for lists and also for overriding keys.

### LINKED ISSUE #4389: Error with `BaseSettings` from a `.env` when trying to load a dictionary in a nested sub-model.
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic  (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

I tried to load a dictionary from a `.env` in a nested sub-model.

Here is my `.env` file:

```
# tests/config/my.env
direct_dict={"v1": "V1", "v2": "V2"}
from_base_model__v1=V1
from_base_model__v2=V2

submodel__direct_dict={"v1": "V1", "v2": "V2"}
submodel__from_base_model__v1=V1
submodel__from_base_model__v2=V2
```

I wrote the code to test reading the dict as a part of the base settings and as a part of the sub-model.
I get the following error:

```
ValidationError: 1 validation error for Settings
submodel -> direct_dict
  value is not a valid dict (type=type_error.dict)
```

If I remove `direct_dict` from the `SubModel` class it works

### Example Code

```Python
import pydantic
from typing import Optional, Dict, Any

class Data(pydantic.BaseModel):
    v1: str
    v2: str

class Submodel(pydantic.BaseModel):
    direct_dict: Dict[str, Any]
    from_base_model: Data

class Settings(pydantic.BaseSettings):
    direct_dict: Dict[str, Any]
    from_base_model: Data
    submodel: Submodel

    class Config:
        case_sensitive = True
        env_file='./tests/config/my.env'
        env_nested_delimiter='__'

s = Settings()
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.9.2
            pydantic compiled: True
                 install path: C:\Users\jlopezcolina\Envs\applications_utils\Lib\site-packages\pydantic
               python version: 3.7.9 (tags/v3.7.9:13c94747c7, Aug 17 2020, 16:30:00) [MSC v.1900 64 bit (AMD64)]  
                     platform: Windows-10-10.0.19041-SP0
     optional deps. installed: ['dotenv', 'typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [X] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [X] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by robertaistleitner**
## Change Summary

Pydantic didn't support setting fields containing "complex values" like dicts or lists from environment variables using nesting. This PR contains an implementation addressing this issue by checking which field the nested environment variable targets to, checking if the field is a complex value and if so, trying to parse the environment value as a JSON.

This already worked if explicitly using `env=...` keyword arg for a field but not for nested environment variables.

## Related issue number

There is no issue created yet.

fix #2304 fix #4389

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by robertaistleitner** [comment]
Please note this PR fixed the same missing feature described in https://github.com/samuelcolvin/pydantic/issues/3971: https://github.com/samuelcolvin/pydantic/pull/4235 and 

What this PR also includes is checking the target field (is it really a complex field) and some more test cases.

File: changes/4216-raistleitner.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Pydantic didn't support setting fields containing "complex values" like dicts or lists from environment variables using nesting. This PR contains an implementation addressing this issue by checking which field the nested environment variable targets to, checking if the field is a complex value and if so, trying to parse the environment value as a JSON.
```
Comment (hramezani) [inline changes/4216-raistleitner.md:1]: I would suggest changing it to a simple sentence.
```suggestion
Add support for setting nested fields containing complex values like dict or list from environment variables
```

File: pydantic/env_settings.py [hunk 2]
```diff
@@ -217,7 +217,22 @@ def field_is_complex(self, field: ModelField) -> Tuple[bool, bool]:
 
         return True, allow_json_failure
 
-    def explode_env_vars(self, field: ModelField, env_vars: Mapping[str, Optional[str]]) -> Dict[str, Any]:
+    @staticmethod
+    def next_field(field: Union[ModelField, None], key: str) -> Union[ModelField, None]:
```
Comment (hramezani) [inline pydantic/env_settings.py:221]:
```suggestion
    def next_field(field: Optional[ModelField], key: str) -> Optional[ModelField]:
```

File: pydantic/env_settings.py [hunk 3]
```diff
@@ -217,7 +217,22 @@ def field_is_complex(self, field: ModelField) -> Tuple[bool, bool]:
...
+        if not field:
+            return None
+
+        if field.sub_fields:
+            # no support for Unions of complex BaseSettings fields
+            return None
```
Comment (hramezani) [inline pydantic/env_settings.py:227]: Maybe we can combine these two conditions:
```suggestion
        if not field or field.sub_fields:
            # no support for Unions of complex BaseSettings fields
            return None
```

File: pydantic/env_settings.py [hunk 4]
```diff
@@ -230,8 +245,25 @@ def explode_env_vars(self, field: ModelField, env_vars: Mapping[str, Optional[st
                 continue
             _, *keys, last_key = env_name.split(self.env_nested_delimiter)
             env_var = result
+            target_field: Union[ModelField, None] = field
```
Comment (hramezani) [inline pydantic/env_settings.py:248]:
```suggestion
            target_field: Optional[ModelField] = field
```

**Comment by hramezani** [comment]
Thanks @robertaistleitner for the patch :+1: 

I left some comments.

Also, I think we can improve the [Parsing environment variable values](https://pydantic-docs.helpmanual.io/usage/settings/#parsing-environment-variable-values) doc and improve the example there with some examples from your change.

**Comment by hramezani** [comment]
please update

**Comment by robertaistleitner** [comment]
> I left some comments.

I worked the comments in as they were because they were all valid as is.
 
> Also, I think we can improve the [Parsing environment variable values](https://pydantic-docs.helpmanual.io/usage/settings/#parsing-environment-variable-values) doc and improve the example there with some examples from your change.

I added a simple example showing the implemented feature.

Please review again.

**Review by hramezani — APPROVED** [review]
Thanks @robertaistleitner for updating. LGTM :+1:

File: pydantic/env_settings.py [hunk 5]
```diff
@@ -217,7 +217,19 @@ def field_is_complex(self, field: ModelField) -> Tuple[bool, bool]:
 
         return True, allow_json_failure
 
-    def explode_env_vars(self, field: ModelField, env_vars: Mapping[str, Optional[str]]) -> Dict[str, Any]:
+    @staticmethod
+    def next_field(field: Optional[ModelField], key: str) -> Optional[ModelField]:
+        if not field or field.sub_fields:
```
Comment (samuelcolvin) [inline pydantic/env_settings.py:222]: This method is only called in one place, are you sure it needs to be a separate function?

If so, please add a docstring explaining what we're doing here.

**Review by samuelcolvin — APPROVED** [review]
Thanks so much for this, I think it looks like a great addition.

My only concern is whether it should be included in V1.10 or wait for V2? (We've got a bit of heat recently for introducing breaking changes in minor and patch updates, hence my paranoia)

There are two potential risks:
* known unknowns: someone has a very weird schema for their settings and a field (e.g. `Union[str, List[str]]`) get's a different value after this PR vs. before
* unknown unknowns: something else unintended changes

I'll think about it, but any input from @robertaistleitner @AngellusMortis @hramezani @PrettyWood on whether this is a suitable change for a minor release would be very welcome.

**Comment by hramezani** [comment]
> My only concern is whether it should be included in V1.10 or wait for V2?

V2 makes more sense to me. The feature itself is great but the implementation is complex and it may introduce breaking changes that is not visible now.

**Comment by PrettyWood** [comment]
I would also differ this for v2

File: pydantic/env_settings.py (hunk 5 above)
Comment (robertaistleitner) [reply pydantic/env_settings.py:222]: This method gets called twice, that's why I moved it to a method:

https://github.com/pydantic/pydantic/pull/4216/files/82f5b971d3a201fea1402eb5ae8b210a03ad5e2a#diff-67b5bf5da243ce3b23618214851e35a3f0951de739f3747800daf6d1273e4962R248

https://github.com/pydantic/pydantic/pull/4216/files/82f5b971d3a201fea1402eb5ae8b210a03ad5e2a#diff-67b5bf5da243ce3b23618214851e35a3f0951de739f3747800daf6d1273e4962R252

Comment (samuelcolvin) [reply pydantic/env_settings.py:222]: Sorry, you're right, my mistake.

Comment (robertaistleitner) [reply pydantic/env_settings.py:222]: I don't think there are collisions with existing setting schemas due to the fact that setting complex values from an environment variable never really worked.

Nevertheless, it's up to you - for now I'm using a monkeypatch that does the job pretty well for me.

Comment (samuelcolvin) [reply pydantic/env_settings.py:222]: I understand where you're coming from, I'm just a bit paranoid at the moment about "breaking" changes, hence deferring this.

**Comment by samuelcolvin** [comment]
Closing since:
1. this is deferred to V2
2. settings logic has moved to [pydantic-settings](https://github.com/pydantic/pydantic-settings)
3. We need to resolve https://github.com/pydantic/pydantic-settings/pull/10 which might solve this.

**Comment by samuelcolvin** [comment]
Thanks so much @robertaistleitner for the proposed change, I'd love you to contribute to pydantic-settings or indeed pydantic in the future.

Sorry we couldn't accept this this time.

**Comment by robertaistleitner** [comment]
@hramezani that's awesome, thanks for your effort!

**Pull request closed without merging**